storage_controller/src/compute_hook.rs

```diff
@@ -71,6 +71,37 @@ impl ComputeHookTenant {
         }
     }
 
+    fn is_sharded(&self) -> bool {
+        matches!(self, ComputeHookTenant::Sharded(_))
+    }
+
+    /// Clear compute hook state for the specified shard.
+    /// Only valid for [`ComputeHookTenant::Sharded`] instances.
+    fn handle_detach(&mut self, tenant_shard_id: TenantShardId, stripe_size: ShardStripeSize) {
+        match self {
+            ComputeHookTenant::Sharded(sharded) => {
+                if sharded.stripe_size != stripe_size
+                    || sharded.shard_count != tenant_shard_id.shard_count
+                {
+                    tracing::warn!("Shard split detected while handling detach")
+                }
+
+                let shard_idx = sharded.shards.iter().position(|(shard_number, _node_id)| {
+                    *shard_number == tenant_shard_id.shard_number
+                });
+
+                if let Some(shard_idx) = shard_idx {
+                    sharded.shards.remove(shard_idx);
```

Comment:
Should this also invalidate the last sent notification?

Reply:
It's an open question. This PR is intended to fix a specific case:
1. Tenant is detached manually by setting its scheduling policy
2. Tenant is re-attached manually by setting its scheduling policy
3. Stale compute notification is sent for the tenant because we did not clear the compute hook state in (1)
4. Compute gets stuck because it can't connect to specified pageservers

The question of how to handle operator driven detaches needs some design work:
* Should we notify the control plane about detaches?
* How do we ensure that a stale notification is never sent out? Some cross reconcile coordination would be required.

Reply:
>Compute gets stuck because it can't connect to specified pageservers

I _think_ this is only a deadlock-ish issue in our tests, because we're updating the compute inline from the controller using the neon_local hook.

For real systems, we would emit these notifications asynchronously, and then when the compute is stuck it will stay stuck until we later send it up to date state for all shards.

So in the test this PR is a fix, and in the field this PR is a hygiene thing, where we will avoid spamming the compute with bogus information across a detach/attach.

Reply:
> For real systems, we would emit these notifications asynchronously, and then when the compute is stuck it will stay stuck until we later send it up to date state for all shards.

I'm not sure we are immune to this in prod. In our tests the compute got stuck in `configuring` state and new notifications were refused. In prod, cplane will respawn the compute after a number of attempts, but we could still produce an availability gap.